module/executiondatasync/pruner/pruner.go

```diff
@@ -0,0 +1,157 @@
+package pruner
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/executiondatasync/tracker"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+	"github.com/rs/zerolog"
+)
+
+const (
+	defaultHeightRangeTarget = uint64(400000)
+	defaultThreshold         = uint64(100000)
+)
+
+type Pruner struct {
+	storage tracker.Storage
+
+	fulfilledHeightsIn    chan<- interface{}
+	fulfilledHeightsOut   <-chan interface{}
+	thresholdChan         chan uint64
+	heightRangeTargetChan chan uint64
+
+	lastFulfilledHeight uint64
+	lastPrunedHeight    uint64
+	heightRangeTarget   uint64
+	threshold           uint64
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataPrunerMetrics
+
+	component.Component
+	cm *component.ComponentManager
+}
+
+type PrunerOption func(*Pruner)
+
+func WithHeightRangeTarget(heightRangeTarget uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.heightRangeTarget = heightRangeTarget
+	}
+}
+
+func WithThreshold(threshold uint64) PrunerOption {
+	return func(p *Pruner) {
+		p.threshold = threshold
+	}
+}
+
+func NewPruner(logger zerolog.Logger, metrics module.ExecutionDataPrunerMetrics, storage tracker.Storage, opts ...PrunerOption) (*Pruner, error) {
+	lastPrunedHeight, err := storage.GetPrunedHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get pruned height: %w", err)
+	}
+
+	fulfilledHeight, err := storage.GetFulfilledHeight()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get fulfilled height: %w", err)
+	}
+
+	fulfilledHeightsIn, fulfilledHeightsOut := util.UnboundedChannel()
```

Comment:
I am wary of using unbounded channels like this. In comparison to unbuffered channels or message stores with observable and capped sizes, it is more difficult to detect bottlenecks or reason about the memory usage of messages awaiting processing. 

The usage here is one where the scope is limited enough to be pretty confident it will work as expected, but I think it's a pattern we should avoid in general on Flow, so I'm inclined to suggest we consider removing it even for this usage. 

Some suggestions for alternatives:
* caller updates mutex-protected values (fulfilled height, threshold, height range) directly, uses `Notifier` to notify pruning thread for every change
* replace `UnboundedChannel` with something similar that buffers only the maximum input in the channel, drop other inputs

Reply:
It's probably not necessary here, since we only really care about processing the most recent item, but I introduced it mainly for other usecases in the requester where we need to process every single item, yet we don't want the producers to be blocked waiting for the consumers to finish processing.

For this specific case, I'm fine replacing it with one of your suggestions.